Type checker doesn't consider trait duality when reasoning about type parameters for associated types

See the testcase here: https://is.gd/WZh43I

I'd normally run this by a few people before filing an issue, but #rust is quiet on a Saturday night. ;-)

@Manishearth thoughts?

Current output:

```
error[E0277]: the trait bound `<D as DomTraversal<<E as TElement>::ConcreteNode>>::BorrowElement: std::borrow::Borrow<E>` is not satisfied
  --> src/main.rs:15:1
   |
11 |   trait DomTraversal<N: TNode> {
   |   ---------------------------- required by `DomTraversal`
...
15 | / fn recalc_style_at<E, D>()
16 | |     where E: TElement,
17 | |           D: DomTraversal<E::ConcreteNode>,
   | |                                            - help: consider further restricting the associated type: `, <D as DomTraversal<<E as TElement>::ConcreteNode>>::BorrowElement: std::borrow::Borrow<E>`
18 | | {
19 | | }
   | |_^ the trait `std::borrow::Borrow<E>` is not implemented for `<D as DomTraversal<<E as TElement>::ConcreteNode>>::BorrowElement`
```

--- a/src/test/ui/associated-types/issue-38917.rs
+++ b/src/test/ui/associated-types/issue-38917.rs
@@ -1,0 +1,25 @@
+// check-pass
+
+use std::borrow::Borrow;
+
+trait TNode: Sized {
+    type ConcreteElement: TElement<ConcreteNode = Self>;
+}
+
+trait TElement: Sized {
+    type ConcreteNode: TNode<ConcreteElement = Self>;
+}
+
+trait DomTraversal<N: TNode> {
+    type BorrowElement: Borrow<N::ConcreteElement>;
+}
+
+#[allow(dead_code)]
+fn recalc_style_at<E, D>()
+where
+    E: TElement,
+    D: DomTraversal<E::ConcreteNode>,
+{
+}
+
+fn main() {}

--- a/src/test/ui/associated-types/issue-40093.rs
+++ b/src/test/ui/associated-types/issue-40093.rs
@@ -1,0 +1,14 @@
+// check-pass
+
+pub trait Test {
+    type Item;
+    type Bundle: From<Self::Item>;
+}
+
+fn fails<T>()
+where
+    T: Test<Item = String>,
+{
+}
+
+fn main() {}

--- a/src/test/ui/associated-types/issue-43475.rs
+++ b/src/test/ui/associated-types/issue-43475.rs
@@ -1,0 +1,10 @@
+// check-pass
+
+trait Foo { type FooT: Foo; }
+impl Foo for () { type FooT = (); }
+trait Bar<T: Foo> { type BarT: Bar<T::FooT>; }
+impl Bar<()> for () { type BarT = (); }
+
+#[allow(dead_code)]
+fn test<C: Bar<()>>() { }
+fn main() { }

--- a/src/test/ui/associated-types/issue-63591.rs
+++ b/src/test/ui/associated-types/issue-63591.rs
@@ -1,0 +1,24 @@
+// check-pass
+
+#![feature(associated_type_bounds)]
+#![feature(type_alias_impl_trait)]
+
+fn main() {}
+
+trait Bar { type Assoc; }
+
+trait Thing {
+    type Out;
+    fn func() -> Self::Out;
+}
+
+struct AssocIsCopy;
+impl Bar for AssocIsCopy { type Assoc = u8; }
+
+impl Thing for AssocIsCopy {
+    type Out = impl Bar<Assoc: Copy>;
+
+    fn func() -> Self::Out {
+        AssocIsCopy
+    }
+}

